core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala

```diff
@@ -0,0 +1,283 @@
+    val storageLevel = event.blockUpdatedInfo.storageLevel
+    val blockId = event.blockUpdatedInfo.blockId.asInstanceOf[RDDBlockId]
+
+    // SPARK-27677. When a block can be fetched from the external shuffle service, the executor can
+    // be removed without hurting the application too much, since the cached data is still
+    // available. So don't count blocks that can be served by the external service.
+    if (storageLevel.isValid && (!fetchFromShuffleSvcEnabled || !storageLevel.useDisk)) {
+      val hadCachedBlocks = exec.cachedBlocks.nonEmpty
+
+      val blocks = exec.cachedBlocks.getOrElseUpdate(blockId.rddId,
+        new mutable.BitSet(blockId.splitIndex))
+      blocks += blockId.splitIndex
+
+      if (!hadCachedBlocks) {
+        exec.updateTimeout()
+      }
+    } else {
+      exec.cachedBlocks.get(blockId.rddId).foreach { blocks =>
+        blocks -= blockId.splitIndex
+        if (blocks.isEmpty) {
+          exec.cachedBlocks -= blockId.rddId
+          if (exec.cachedBlocks.isEmpty) {
+            exec.updateTimeout()
+          }
+        }
+      }
+    }
+  }
+
+  override def onUnpersistRDD(event: SparkListenerUnpersistRDD): Unit = {
+    executors.values().asScala.foreach { exec =>
+      exec.cachedBlocks -= event.rddId
+      if (exec.cachedBlocks.isEmpty) {
+        exec.updateTimeout()
+      }
+    }
+  }
+
+  // Visible for testing.
+  private[dynalloc] def isExecutorIdle(id: String): Boolean = {
+    Option(executors.get(id)).map(_.isIdle).getOrElse(throw new NoSuchElementException(id))
+  }
+
+  // Visible for testing
+  private[dynalloc] def timedOutExecutors(when: Long): Seq[String] = {
+    executors.asScala.flatMap { case (id, tracker) =>
+      if (tracker.timeoutAt <= when) Some(id) else None
+    }.toSeq
+  }
+
+  // Visible for testing
+  def executorsPendingToRemove(): Set[String] = {
+    executors.asScala.filter { case (_, exec) => exec.pendingRemoval }.keys.toSet
+  }
+
+  /**
+   * This method should be used when updating executor state. It guards against a race condition in
+   * which the `SparkListenerTaskStart` event is posted before the `SparkListenerBlockManagerAdded`
+   * event, which is possible because these events are posted in different threads. (see SPARK-4951)
+   */
+  private def ensureExecutorIsTracked(id: String): Tracker = {
+    executors.computeIfAbsent(id, _ => new Tracker())
+  }
+
+  private def updateNextTimeout(newValue: Long): Unit = {
+    while (true) {
+      val current = nextTimeout.get()
+      if (newValue >= current || nextTimeout.compareAndSet(current, newValue)) {
+        return
+      }
+    }
+  }
+
+  private class Tracker {
+    @volatile var timeoutAt: Long = Long.MaxValue
+
+    // Tracks whether this executor is thought to be timed out. It's used to detect when the list
+    // of timed out executors needs to be updated due to the executor's state changing.
+    @volatile var timedOut: Boolean = false
+
+    var pendingRemoval: Boolean = false
```

Comment:
I think this should be volatile too -- its accessed by the EAM thread and the listenerbus thread

Reply:
I don't think it needs to. I tried very hard to avoid volatile when not necessary; in this code, it's generally only needed when ordering is important when checking multiple variables. That is not the case for this one, since it doesn't really depend on any other variables.

At first I also didn't have `timedOut` and `timeoutAt` as volatiles, but I couldn't 100% convince me that it was not needed. But in this case I'm pretty sure it doesn't do anything useful.

Reply:
ok, I guess the one access which is in the listener thread is in `onExecutorRemoved()`.  If that sees the stale value, then it will incorrectly set the `nextTimeout`, telling the monitor to recompute the list.  But while that is inefficient, its never incorrect.  And the hope is, its rare enough that its still better to avoid making it volatile.

is that right?

I'd say it at least deserves a comment in onExecutorRemoved.

Reply:
>  If that sees the stale value, then it will incorrectly set the nextTimeout

Right, but notice that isn't really related to the variable being volatile or not, it's just a race that can also occur if the variable is volatile.

Writing to that variable is an atomic operation. Volatile controls what happens to other reads / writes around that volatile write. That's why it's needed when the order of the reads and writes is important. But here it isn't.